apps/trading/lib/hooks/use-position-closeout-notification.tsx

```diff
@@ -0,0 +1,327 @@
+      // hides toast instead of removing is so it won't be re-added on rerender
+      updateToast(createToastId(tradeId), { hidden: true });
+      setSeen(tradeId);
+    },
+    [setSeen, updateToast]
+  );
+
+  const [errorCount, setErrorCount] = useState(0);
+
+  const onRejection = useCallback(
+    () => setErrorCount((errorCount) => errorCount + 1),
+    [setErrorCount]
+  );
+
+  const pushNotification = useCallback(
+    (trade: TradeFieldsFragment | TradeUpdateFieldsFragment) => {
+      const marketId =
+        trade.__typename === 'Trade'
+          ? trade.market.id
+          : trade.__typename === 'TradeUpdate'
+          ? trade.marketId
+          : undefined;
+      if (!marketId) return;
+
+      const market = markets?.[marketId];
+
+      // The pubKey associated with closed out position
+      let pubKey: string | undefined = undefined;
+      switch (trade.aggressor) {
+        case Side.SIDE_SELL:
+          if (trade.__typename === 'Trade') pubKey = trade.buyer.id;
+          if (trade.__typename === 'TradeUpdate') pubKey = trade.buyerId;
+          break;
+        case Side.SIDE_BUY:
+          if (trade.__typename === 'Trade') pubKey = trade.seller.id;
+          if (trade.__typename === 'TradeUpdate') pubKey = trade.sellerId;
+          break;
+        default:
+        case Side.SIDE_UNSPECIFIED:
+          pubKey = undefined;
+          break;
+      }
+
+      const toastId = createToastId(trade.id);
+
+      if (!market || !pubKey || seen.includes(trade.id) || hasToast(toastId)) {
+        return;
+      }
+
+      const closeOutPrice = addDecimalsFormatNumber(
+        trade.price,
+        market.decimalPlaces
+      );
+
+      let asset: Pick<AssetFieldsFragment, 'symbol'> | undefined = undefined;
+      if (
+        market.tradableInstrument.instrument.product.__typename === 'Future' ||
+        market.tradableInstrument.instrument.product.__typename === 'Perpetual'
+      ) {
+        asset = market.tradableInstrument.instrument.product.settlementAsset;
+      } else if (
+        market.tradableInstrument.instrument.product.__typename === 'Spot'
+      ) {
+        asset = market.tradableInstrument.instrument.product.quoteAsset;
+      }
+
+      if (!asset) return;
+
+      const toast: Toast = {
+        id: createToastId(trade.id),
+        intent: Intent.Warning,
+        onClose: onClose(trade.id),
+        content: (
+          <PositionClosedOutToastContent
+            closeOutPrice={closeOutPrice}
+            asset={asset}
+            marketId={market.id}
+            partyId={pubKey}
+            onClose={onClose(trade.id)}
+          />
+        ),
+      };
+      setToast(toast);
+    },
+    [hasToast, markets, onClose, seen, setToast]
+  );
+
+  useEffect(() => {
+    let subscription: { unsubscribe: () => void } | undefined = undefined;
+    let ignore = false;
+
+    if (!pubKeys) {
+      return;
+    }
+
+    const runQueries = () => {
+      if (ignore) return;
```

Comment:
The ignore needs to be used inside the async callbacks to properly ignore 'stale' requests. This is another reason to just use the generated hook, which should handle race conditions if the fetch is invoked twice rapidly